src/Tests/Microsoft.DotNet.PackageInstall.Tests/ToolPackageDownloaderTests.cs

```diff
@@ -443,70 +322,142 @@ public void GivenAEmptySourceAndNugetConfigInstallSucceeds(bool testMockBehavior
 
-            static void FailedStepAfterSuccessRestore() => throw new GracefulException("simulated error");
+            Action a = () =>
+            {
+                using (var t = new TransactionScope(
+                    TransactionScopeOption.Required,
+                    TimeSpan.Zero))
+                {
+                    Action first = () => downloader.InstallPackage(new PackageLocation(additionalFeeds: new[] { source }),
+                        packageId: TestPackageId,
+                        verbosity: TestVerbosity,
+                        versionRange: VersionRange.Parse(TestPackageVersion),
+                        targetFramework: _testTargetframework,
+                        isGlobalTool: true);
+
+                    first.Should().NotThrow();
+
+                    downloader.InstallPackage(new PackageLocation(additionalFeeds: new[] { source }),
+                        packageId: TestPackageId,
+                        verbosity: TestVerbosity,
+                        versionRange: VersionRange.Parse(TestPackageVersion),
+                        targetFramework: _testTargetframework,
+                        isGlobalTool: true);
+
+                    t.Complete();
+                }
+            };
+
+            a.Should().Throw<ToolPackageException>().Where(
+                ex => ex.Message ==
+                      string.Format(
+                          CommonLocalizableStrings.ToolPackageConflictPackageId,
+                          TestPackageId,
+                          TestPackageVersion));
+
+            AssertInstallRollBack(fileSystem, store);
+        }
+
+        [Theory]
+        [InlineData(false)]
+        [InlineData(true)]
+        public void GivenFailureWhenInstallLocalToolsItWillRollbackPackageVersion(bool testMockBehaviorIsInSync)
+        {
+            var source = GetTestLocalFeedPath();
+
+            var (store, storeQuery, downloader, uninstaller, reporter, fileSystem) = Setup(
+                useMock: testMockBehaviorIsInSync,
+                feeds: GetMockFeedsForSource(source));
+
+            static void FailedStepAfterSuccessDownload() => throw new GracefulException("simulated error");
 
             Action a = () =>
             {
                 using (var t = new TransactionScope(
                     TransactionScopeOption.Required,
                     TimeSpan.Zero))
                 {
-                    installer.InstallPackage(new PackageLocation(additionalFeeds: new[] { source }),
+                    downloader.InstallPackage(new PackageLocation(additionalFeeds: new[] { source }),
                         packageId: TestPackageId,
+                        verbosity: TestVerbosity,
                         versionRange: VersionRange.Parse(TestPackageVersion),
                         targetFramework: _testTargetframework);
 
-                    FailedStepAfterSuccessRestore();
+                    FailedStepAfterSuccessDownload();
```

Comment:
I think it would be a good idea to assert that `localToolDownloadDir` exists before throwing the exception.  That would help us be sure that we're checking the right path.